doc/man7/openssl-threads.pod

```diff
@@ -0,0 +1,66 @@
+=pod
+
+=head1 NAME
+
+openssl-threads - Using multiple threads with OpenSSL
+
+=head1 DESCRIPTION
+
+OpenSSL can be built with or without threads support. The most important
+use of this support is so that OpenSSL itself can use a single consistent
+API, as shown in L<CRYPTO_THREAD_run_once(3)/EXAMPLES>.
+Multi-platform applications can also use this API.
+
+In particular, being configured for threads-support does not imply that
+all objects can be used simultaneously in more than one thread.
+To repeat: I<most objects are not safe for simultaneous use>.
+Exceptions to this should be documented on the specific manual pages, and
+some general high-level guidance is given here.
+
+One major use of the OpenSSL thread API is to implement reference counting.
+Many object within OpenSSL are reference-counted, so resources are not
+released, until the last reference is removed.
+References are often increased automatically (such as when an B<X509>
+certificate object is added into an B<X509_STORE> trust store).
+There is often an I<object>_up_ref() API that can be used to increase the
+reference.
+Failure to match I<object>_up_ref() calls with the right number of
+I<object>_free() calls is a common source of memory leaks when a program
+exits.
+The various C<set1> and C<get1> functions associated with an object are
+used to set or get attributes of an object; they increase the refernece
+count.
+The C<set0> and C<get0> functions are similar, but do not increase the
+reference count, passing "ownership" appropriately.
+Functions with a simple C<set> or C<get> are historical; and the documentation
+must state how the references are handled.
+
+Objects can often be safely used by multiple threads simultaneously
+as long as the API's being invoked don't modify the object; in this
+case the parameter is usually marked in the API as C<const>.
+Because of the implementation (and because C does not have the C++
+concept of C<mutable const>) not all applicable API's have the C<const>
+modifier.
+Another instance of safe use is when an object's internal state,
+such as cached values, are done with locks.
+One example of this is the reference counting API's described above.
```

Comment:
Should we be more explicit and say that const doesn't mean immutable or physical constant, but does mean logical constant and thread safe? It's then up to the object itself to take a lock if it wants to modify something, but remain logically the same. For instance X509_cmp() currently takes 2 `const X509 *`, but will internally remove the const to update cached values. It will internally take a lock to update that cache, and so it's actually thread safe, and logically constant. X509_check_purpose() has a comment that it also wants const but can't because it wants to update the cache, and so could actually also have the const.

I would remove the "mutable const" part, it's not really relevant.

Reply:
I tried to add some text about this, in the commit I am about to push.

Reply:
The "Another instance", is really the same as the X509_cmp() example, cache something.

Note that none of the up_ref() calls currently take a const because they modify the counter, but as documented here, they could and probably should.

Reply:
I couldn't think of another type of thing.  But it's "another instance" because it might not be obvious to developers.  If you can think of something that is different, let me know.  If you think I should remove that other instance (I would rather not because of the obviousness I mentioned), let me knwo.

I created #13796 for the "const" in upref/downref.

Reply:
How is a user supposed to use the information provided in this paragraph? Really, I think you are saying that there are some functions that mutate objects that may still be thread safe. We should seek to document what those functions are (doesn't have to be in this PR)

Reply:
It's a warning that the behavior, if not otherwise documented, is undefined.  "Here be dragons."